### PR TITLE
Change certain replica failures not to fail the replica shard

### PR DESCRIPTION
**Pull request by dakrone**
This changes the way that replica failures are handled such that not all
failures will cause the replica shard to be failed or marked as stale.

In some cases such as refresh operations, or global checkpoint syncs, it is
"okay" for the operation to fail without the shard being failed (because no data
is out of sync). In these cases, instead of failing the shard we should simply
fail the operation, and, in the event it is a user-facing operation, return a
5xx response code including the shard-specific failures.

This was accomplished by having two forms of the `Replicas` proxy, one that is
for non-write operations that does not fail the shard, and one that is for write
operations that will fail the shard when an operation fails.

Relates to #10708

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java [hunk 1]
```diff
@@ -209,14 +209,20 @@ public void onFailure(Exception replicaException) {
...
+                    // TODO: Boaz wanted to get rid of this warning, should we?
                     logger.warn(
                         (org.apache.logging.log4j.util.Supplier<?>)
                             () -> new ParameterizedMessage("[{}] {}", shard.shardId(), message), replicaException);
-                    replicasProxy.failShard(shard, replicaRequest.primaryTerm(), message, replicaException,
-                        ReplicationOperation.this::decPendingAndFinishIfNeeded,
-                        ReplicationOperation.this::onPrimaryDemoted,
-                        throwable -> decPendingAndFinishIfNeeded()
-                    );
+                    final boolean shardWasFailed = replicasProxy.failShardIfNeeded(shard, replicaRequest.primaryTerm(), message,
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java]: I'd rather not introduce more code paths. I would vote to always have the callback pattern. If people have nothing to do they can call onSuccess on the same thread.

File: core/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java [hunk 2]
```diff
@@ -209,14 +209,20 @@ public void onFailure(Exception replicaException) {
                     shardReplicaFailures.add(new ReplicationResponse.ShardInfo.Failure(
                         shard.shardId(), shard.currentNodeId(), replicaException, restStatus, false));
                     String message = String.format(Locale.ROOT, "failed to perform %s on replica %s", opType, shard);
+                    // TODO: Boaz wanted to get rid of this warning, should we?
                     logger.warn(
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java]: we plan to use this for background operations like the global check point sync, where it's not a big deal if this happen. I don't want the logs to have warning in them. Instead, implementations (i.e., TransportWriteAction) of `failShardIfNeeded` can log a warning if they are going to fail the shard

File: core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java [hunk 3]
```diff
@@ -1030,8 +1034,22 @@ public String allocationId() {
         }
     }
 
+    /**
+     * The {@code ReplicasProxy} is an implementation of the {@code Replicas}
+     * interface that performs the actual {@code ReplicaRequest} on the replica
+     * shards. It also encapsulates the logic required for failing the replica
+     * if deemed necessary as well as marking it as stale when needed.
+     */
     final class ReplicasProxy implements ReplicationOperation.Replicas<ReplicaRequest> {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java]: why make this final? WriteReplicasProxy can share a lot of it's logic with this one?

File: core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java [hunk 4]
```diff
@@ -1030,8 +1034,22 @@ public String allocationId() {
...
 
+        private final ClusterService clusterService;
+        private final TransportReplicationAction replicationAction;
+
+        ReplicasProxy(ClusterService clusterService, TransportReplicationAction replicationAction) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java]: this is an inner class anyway, why pass these along? If you want to make it a static inner class, I'm good with it, although I'm not sure if it's worth the extra clutter.

File: core/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java [hunk 5]
```diff
@@ -364,24 +376,48 @@ private void finishAsFailed(Exception exception) {
...
+         * @param allocationId     allocation id to remove from the set of in-sync allocation ids
+         * @param primaryTerm      the primary term of the primary shard when requesting the failure
+         * @param onSuccess        a callback to call when the allocation id has been successfully removed from the in-sync set.
+         * @param onPrimaryDemoted a callback to call when the request failed because the current primary was already demoted
+         *                         by the master.
+         * @param onIgnoredFailure a callback to call when the request failed, but the failure can be safely ignored.
          */
-        void failShard(ShardRouting replica, long primaryTerm, String message, Exception exception, Runnable onSuccess,
-                       Consumer<Exception> onPrimaryDemoted, Consumer<Exception> onIgnoredFailure);
+        void markShardCopyAsStaleIfNeeded(ShardId shardId, String allocationId, long primaryTerm, Runnable onSuccess,
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java]: why do we need two of these? can't we have just markShardCopyAsStaleIfNeeded?  we can rename `markUnavailableShardsAsStale` to `markUnavailableShardsAsStaleIfNeeded`

File: core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java [hunk 6]
```diff
@@ -1041,26 +1059,37 @@ public void performOn(ShardRouting replica, ReplicaRequest request, ActionListen
...
+                                                 Consumer<Exception> onPrimaryDemoted, Consumer<Exception> onIgnoredFailure) {
+            // Just like with failing the shard, it should only be marked as
+            // stale when write operations fail, so for regular operations
+            // don't mark it as stale when not necessary.
         }
 
         @Override
         public void markShardCopyAsStale(ShardId shardId, String allocationId, long primaryTerm, Runnable onSuccess,
                                          Consumer<Exception> onPrimaryDemoted, Consumer<Exception> onIgnoredFailure) {
             shardStateAction.remoteShardFailed(shardId, allocationId, primaryTerm, "mark copy as stale", null,
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java]: this logic belongs in transportWriteAction

File: core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointSyncAction.java [hunk 7]
```diff
@@ -43,8 +43,8 @@
 
 import java.io.IOException;
 
-public class GlobalCheckpointSyncAction extends TransportReplicationAction<GlobalCheckpointSyncAction.PrimaryRequest,
-    GlobalCheckpointSyncAction.ReplicaRequest, ReplicationResponse> {
+public class GlobalCheckpointSyncAction extends TransportReplicationAction<GlobalCheckpointSyncAction.GlobalCheckpointPrimaryRequest,
+    GlobalCheckpointSyncAction.GlobalCheckpointReplicaRequest, ReplicationResponse> {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointSyncAction.java]: why rename this , it's already scoped under GlobalCheckpointSyncAction. it gets so long...

**Review by bleskes — CHANGES_REQUESTED** [review]
Thx @dakrone. I left a bunch of initial comments. Maybe I miss something obvious but where is the logic to respond with an error if one of the shard copies fail (except in the TransportWriteAction)?

File: core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointSyncAction.java (hunk 7 above)
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointSyncAction.java]: I pushed 68d0ddd456d5e42db40c06261af8e8a19c201533 for this.

**Comment by dakrone** [comment]
>  Maybe I miss something obvious but where is the logic to respond with an error if one of the shard copies fail (except in the TransportWriteAction)?

That logic is in the BroadcastShardResponse now, (`getStatus()`), which is used on the REST side. The error reporting was already there, this just changes the response to be a 500 instead of 200 in the event there is an error.

File: core/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java (hunk 5 above)
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java]: We need two because there are situations when a replica shard needs to be marked as stale without a write action, I originally had this and then the tests that restart ES exposed that they still need to be able to mark it as stale without any actions.

File: core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java (hunk 3 above)
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java]: I originally had them sharing knowledge (subclassing), however, `ReplicasProxy` used parts of `TransportReplicationAction` and made `WriteActionReplicasProxy` not able to subclass it well.

File: core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java (hunk 4 above)
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java]: I originally tried to factor this out into a non-inner class, but turned out that the generics prevented it, so I will removed these.

Comment (dakrone) [inline core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java]: I pushed b5e91badc401a5aab1f20209ce0096881b07992f

File: core/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java (hunk 2 above)
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java]: Okay, I pushed ccda991ee569490ce2ced3f5bff7763e867e20ac

File: core/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java (hunk 1 above)
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java]: Okay, I pushed 3589d9c39d7030b9d7817764626520cb976ba9b2 that switches this to use the callback

File: core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java (hunk 6 above)
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java]: As I mentioned in my previous comment, we still need a way to mark the replica as stale when the shard is unavailable.

What we don't need is the "if needed" version, because we only do this when the shard is not available, not when an operation fails.

I've removed the "if needed" version of this in 1f58744362b756bc9ef6f6bfb412c2836890d8f7

**Comment by dakrone** [comment]
Thanks @bleskes, I think I pushed commits and added comments for all of your feedback so far.

File: core/src/main/java/org/elasticsearch/action/support/broadcast/BroadcastResponse.java [hunk 8]
```diff
@@ -71,6 +74,17 @@ public int getFailedShards() {
     }
 
     /**
+     * The REST status that should be used for the response
+     */
+    public RestStatus getStatus() {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/broadcast/BroadcastResponse.java]: Since this is effects all response that inherit from BroadcastResponse - can you check that all of them do the right thing and filter shard not available failures? might be good to add an assertion in the constructors, looking at all the exceptions this get.

File: core/src/main/java/org/elasticsearch/action/support/broadcast/BroadcastResponse.java [hunk 9]
```diff
@@ -71,6 +74,17 @@ public int getFailedShards() {
...
+        if (failedShards > 0) {
+            return RestStatus.INTERNAL_SERVER_ERROR;
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/broadcast/BroadcastResponse.java]: shall we make an heroic effort to take the worst of all underlying rest status? maybe just the first ?

File: core/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java [hunk 10]
```diff
@@ -314,10 +309,13 @@ private void finishAsFailed(Exception exception) {
...
     }
 
+    /**
+     * An encapsulation of an operation that is to be performed on the primary shard
+     */
     public interface Primary<
-                Request extends ReplicationRequest<Request>,
-                ReplicaRequest extends ReplicationRequest<ReplicaRequest>,
-                PrimaryResultT extends PrimaryResult<ReplicaRequest>
+                R extends ReplicationRequest<R>,
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java]: nit: Why do we call this R and use an T suffix on the rest?

File: core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java [hunk 11]
```diff
@@ -368,9 +372,9 @@ public void onFailure(Exception e) {
         protected ReplicationOperation<Request, ReplicaRequest, PrimaryResult<ReplicaRequest, Response>> createReplicatedOperation(
             Request request, ActionListener<PrimaryResult<ReplicaRequest, Response>> listener,
             PrimaryShardReference primaryShardReference, boolean executeOnReplicas) {
-            return new ReplicationOperation<>(request, primaryShardReference, listener,
-                executeOnReplicas, replicasProxy, clusterService::state, logger, actionName
-            );
+            return new ReplicationOperation<Request, ReplicaRequest, PrimaryResult<ReplicaRequest, Response>>(request,
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java]: out of curiosity - why does <> not work?

File: core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java (hunk 3 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java]: This seems to work fine for me?

```
diff --git a/core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java b/core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java
index db037ad..2cd1d8e 100644
--- a/core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java
+++ b/core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java
@@ -1040,7 +1040,7 @@ public abstract class TransportReplicationAction<
      * shards. It also encapsulates the logic required for failing the replica
      * if deemed necessary as well as marking it as stale when needed.
      */
-    final class ReplicasProxy implements ReplicationOperation.Replicas<ReplicaRequest> {
+    class ReplicasProxy implements ReplicationOperation.Replicas<ReplicaRequest> {
 
         @Override
         public void performOn(ShardRouting replica, ReplicaRequest request, ActionListener<ReplicationOperation.ReplicaResponse> listener) {
diff --git a/core/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java b/core/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java
index ac79a8c..8a168ca 100644
--- a/core/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java
+++ b/core/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java
@@ -74,7 +74,7 @@ public abstract class TransportWriteAction<
 
     @Override
     protected ReplicationOperation.Replicas newReplicasProxy() {
-        return new WriteActionReplicasProxy(shardStateAction);
+        return new WriteActionReplicasProxy();
     }
 
     /**
@@ -335,26 +335,7 @@ public abstract class TransportWriteAction<
      * replicas, where a failure to execute the operation should fail
      * the replica shard and/or mark the replica as stale.
      */
-    final class WriteActionReplicasProxy implements ReplicationOperation.Replicas<ReplicaRequest> {
-
-        private final ShardStateAction shardStateAction;
-
-        WriteActionReplicasProxy(ShardStateAction shardStateAction) {
-            this.shardStateAction = shardStateAction;
-        }
-
-        @Override
-        public void performOn(ShardRouting replica, ReplicaRequest request, ActionListener<ReplicationOperation.ReplicaResponse> listener) {
-            String nodeId = replica.currentNodeId();
-            final DiscoveryNode node = clusterService.state().nodes().get(nodeId);
-            if (node == null) {
-                listener.onFailure(new NoNodeAvailableException("unknown node [" + nodeId + "]"));
-                return;
-            }
-            final ConcreteShardRequest<ReplicaRequest> concreteShardRequest =
-                    new ConcreteShardRequest<>(request, replica.allocationId().getId());
-            sendReplicaRequest(concreteShardRequest, node, listener);
-        }
+    class WriteActionReplicasProxy extends ReplicasProxy {
 
         @Override
         public void failShardIfNeeded(ShardRouting replica, long primaryTerm, String message, Exception exception,
```

File: core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java (hunk 6 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java]: > We need two because there are situations when a replica shard needs to be marked as stale without a write action

Which situations do you mean?

File: core/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java [hunk 12]
```diff
@@ -124,7 +138,11 @@ public synchronized void respond(ActionListener<Response> listener) {
         protected void respondIfPossible(Exception ex) {
             if (finishedAsyncActions && listener != null) {
                 if (ex == null) {
-                    super.respond(listener);
+                    if (finalResponseIfSuccessful != null) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java]: why this change?

File: core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointSyncAction.java [hunk 13]
```diff
@@ -129,7 +129,7 @@ public String toString() {
         private ReplicaRequest() {
         }
 
-        public ReplicaRequest(PrimaryRequest primaryRequest, long checkpoint) {
+        public ReplicaRequest(ReplicationRequest primaryRequest, long checkpoint) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointSyncAction.java]: Why this change?

File: core/src/main/java/org/elasticsearch/rest/action/RestFieldStatsAction.java [hunk 14]
```diff
@@ -98,7 +98,7 @@ public RestResponse buildResponse(FieldStatsResponse response, XContentBuilder b
                     builder.endObject();
                 }
                 builder.endObject();
-                return new BytesRestResponse(RestStatus.OK, builder);
+                return new BytesRestResponse(response.getStatus(), builder);
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/rest/action/RestFieldStatsAction.java]: these are not related changes - can we maybe move them to a separate PR?

**Review by bleskes — CHANGES_REQUESTED** [review]
Thx @dakrone . I responded.

File: core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java (hunk 3 above)
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java]: @bleskes that's actually a scary thing, that "*works*" for Intellij's compiler (it reports no compilation errors), but if you actually run `gradle compileJava`:

```
/home/hinmanm/es/elasticsearch/core/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java:347: error: name clash: failShardIfNeeded(ShardRouting,long,String,Exception,Runnable,Consumer<Exception>,Consumer<Exception>) in TransportWriteAction.WriteActionReplicasProxy and failShardIfNeeded(ShardRouting,long,String,Exception,Runnable,Consumer<Exception>,Consumer<Exception>) in TransportReplicationAction.ReplicasProxy have the same erasure, yet neither overrides the other
        public void failShardIfNeeded(ShardRouting replica, long primaryTerm, String message, Exception exception,
                    ^
/home/hinmanm/es/elasticsearch/core/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java:346: error: method does not override or implement a method from a supertype
        @Override
        ^
```

It fails because of type erasure (I think due to the crazy generics that are in use), I tried a lot of different ways to get around it but I wasn't able to find a way.

File: core/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java (hunk 10 above)
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java]: I have seen two standard "formats", one is the single char format, which the JDK uses a lot like `Collection<E>`, so that's why the `R`. For the others, `PrimaryResult` as `PR` I could do, but I figured since it wasn't a single char I'd suffix it with `T` for type (another format I've seen in codebases).

I'm happy to change it to `R`, and `PR` if you think that would be clearer

File: core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java (hunk 11 above)
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java]: This was added unnecessarily, I've removed it in 5eafb1866e85b994fafd178d72c873bba9838e71

File: core/src/main/java/org/elasticsearch/action/support/broadcast/BroadcastResponse.java (hunk 8 above)
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/action/support/broadcast/BroadcastResponse.java]: Certainly, I added an assert in f79c0f724819ff428c7c2c1d9bc2701521d96a81

File: core/src/main/java/org/elasticsearch/action/support/broadcast/BroadcastResponse.java (hunk 9 above)
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/action/support/broadcast/BroadcastResponse.java]: > shall we make an heroic effort to take the worst of all underlying rest status? maybe just the first ?

I don't know, how do you define "worst"? Is a 403 worse than a 500 or is a 500 worse? We can totally go with first though, if that's desired.

File: core/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java (hunk 12 above)
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java]: This was a leftover from me trying something out, I've removed it in 56cc63f612e5281555d5da1202dc9a44748eceea

File: core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointSyncAction.java (hunk 13 above)
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointSyncAction.java]: I manually reverted the changes to GlobalCheckpointSyncAction and I missed one :)

I undid it in f7b29d78395c2e13c715f6c167d6e8cbffe28e6e

File: core/src/main/java/org/elasticsearch/rest/action/RestFieldStatsAction.java (hunk 14 above)
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/rest/action/RestFieldStatsAction.java]: Sure, though I think it makes it more consistent to change these to be the same, I'll open a separate PR for it after this one to change them to be 500 if a shard fails and remove all of these except for the refresh action

Comment (dakrone) [inline core/src/main/java/org/elasticsearch/rest/action/RestFieldStatsAction.java]: I pushed d16441b19d606d87166d82345c1197f7fa6d8386 for this

**Comment by dakrone** [comment]
@bleskes I think I addressed all the feedback for this round, thanks again for taking a look!

File: core/src/main/java/org/elasticsearch/action/support/broadcast/BroadcastResponse.java (hunk 9 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/broadcast/BroadcastResponse.java]: I think 5xx (server error) is worst than 4xx, but let's just go with the first.

File: core/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java (hunk 10 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java]: I would go with a T suffix - i.e., RequestT

File: core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java [hunk 15]
```diff
@@ -99,15 +99,15 @@
...
     private final String executor;
 
     // package private for testing
     private final String transportReplicaAction;
     private final String transportPrimaryAction;
-    private final ReplicasProxy replicasProxy;
+    private final ReplicationOperation.Replicas replicasProxy;
 
     protected TransportReplicationAction(Settings settings, String actionName, TransportService transportService,
                                          ClusterService clusterService, IndicesService indicesService,
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java]: we can remove the shard state action here, no?

File: core/src/test/java/org/elasticsearch/action/support/replication/TransportWriteActionTests.java [hunk 16]
```diff
@@ -222,6 +402,80 @@ protected TestResponse newResponseInstance() {
...
+            final ShardId shardId = new ShardId(indexMetaData.getIndex(), shard);
+            if (shard > indexMetaData.getNumberOfShards()) {
+                throw new ShardNotFoundException(shardId);
+            }
+            return mockIndexShard(shardId, clusterService);
+        });
+        return indexService;
+    }
+
+    final IndicesService mockIndicesService(ClusterService clusterService) {
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/action/support/replication/TransportWriteActionTests.java]: can we share this and mockIndexShard with TransportReplicationAction? is there any point in that?

**Review by bleskes — APPROVED** [review]
LGTM. Left some comments that you can adopt or not. No need for another review. Thx @dakrone

**Comment by dakrone** [comment]
Thank you very much for reviewing this @bleskes!
